p2p/discovery/broadcast.go

```diff
@@ -0,0 +1,187 @@
+package discovery
+
+import (
+	"context"
+	"math/rand"
+	"net"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/libp2p/go-libp2p-host"
+	"github.com/libp2p/go-libp2p-peer"
+	pstore "github.com/libp2p/go-libp2p-peerstore"
+	ma "github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr-net"
+)
+
+type broadcastService struct {
+	host          host.Host
+	listenSock    *net.UDPConn   // Socket listening for broadcasts
+	sendSocks     []*net.UDPConn // Sockets that send broadcasts
+	lk            sync.Mutex
+	notifees      []Notifee
+	broadcastPort int
+	active        bool
+}
+
+//
+// Bind a socket to send and recv broadcasts
+//
+func bindBroadcast(ip net.IP, port int) (*net.UDPConn, error) {
+	conn, err := net.ListenUDP("udp4", &net.UDPAddr{IP: ip, Port: port})
+	if err == nil {
```

Comment:
The go idiom is to (almost) always write:

```go
if err != nil {
    return nil, err
}
```

That helps prevent nesting/confusing control flow.

Reply:
I would argue that typing the same early return idiom over and over and over is worse for understanding control flow.

Reply:
It's just how go is written. However, I'd disagree:

1. It makes it very easy to recognize where errors are being returned.
2. It reduces indentation.
3. It's consistent with all other go code.

Now, one can argue that there should be a *better* way to return errors in go but that's a separate issue (and not something we can do anything about).